### DISCUSSION TITLE
opening a network / integration questions

### DISCUSSION THREAD

**Discussion by bsenthilr** [microsoft/CCF#6165] [Q&A] [answered]
Trying to add a activate a member then open the network. Notice lots of changes from 1.x.
Assuming its still possible to open network without any user added.

Following samples (4.x) https://microsoft.github.io/CCF/release/4.x/governance/open_network.html

The Json body does not look right. The transition_service_to_open action does not have any arguments. But removing args, was able to create a proposal for transition_service_to_open.

Reference to trying to open with service identity.
```
cat transition_service_to_open.json  | jq
{
  "actions": [
    {
      "name": "transition_service_to_open",
      "args": {
        "next_service_identity": "-----BEGIN CERTIFICATE-----\nMIIBvTCCAUOgAwIBAgIQYKj4cvVmpOc1P0YygYcWNTAKBggqhkjOPQQDAzAWMRQw\nEgYDVQQDDAtDQ0YgTmV0d29yazAeFw0yNDA1MDMwMjMwMjhaFw0yNTA1MDMwMjMw\nMjdaMBYxFDASBgNVBAMMC0NDRiBOZXR3b3JrMHYwEAYHKoZIzj0CAQYFK4EEACID\nYgAEDUd1iUbso2CGJBvfzVAc8U3qnymDZkMyB8kCeqqIyT4d8IsH1M7wao6VxY4a\nF80CpTY/cebd3mqTwrkeY3WFfhdNOR7kUts210OeBmTDRjTm7A6LVb2Of6vd56Y2\nxRq1o1YwVDASBgNVHRMBAf8ECDAGAQH/AgEAMB0GA1UdDgQWBBRl59w+hk1VWsfu\nObYDoFrnZOdQ4jAfBgNVHSMEGDAWgBRl59w+hk1VWsfuObYDoFrnZOdQ4jAKBggq\nhkjOPQQDAwNoADBlAjB+6vze9jepzB2NuQhKl2IJsaYI1c+HMrqT1ieyLfJ2R795\nitcYk/3UGJtKxP+JZxcCMQC8yFDGbEokQv/5Lt0x5gkeuj5zJCc/1MDt7mkG+RT7\nbJwekZOIi7yxf6lry1mljgQ=\n-----END CERTIFICATE-----\n"
      }
    }
  ]
}
root@ccf-governor:/tfl/dev/opt/governor/ccf/bin# ./ccf_cose_sign1 --ccf-gov-msg-type proposal --ccf-gov-msg-created_at `date -uIs` --signing-cert ../keys/memberccf_cert.pem --signing-key ../keys/memberccf_privk.pem --content transition_service_to_open.json | curl https://192.168.122.82:443/gov/proposals --cacert ../run/service_cert.pem --data-binary @- -H "content-type: application/cose"
{"error":{"code":"ProposalFailedToValidate","message":"Proposal failed to validate: transition_service_to_open at position 0 failed validation: Error: Proposal does not accept any argument, found \"[object Object]\"\n    at checkNone (public:ccf.gov.constitution[0]:73)\n    at <anonymous> (public:ccf.gov.constitution[0]:452)\n    at validate (public:ccf.gov.constitution[0]:796)\n"}}
```

The next step following the documentation adjusting to my proposal_id etc., to cast a vote following the doc:
```
$ cat vote_accept.json
{
    "ballot": "export function vote (proposal, proposerId) { return true }"
}

./ccf_cose_sign1 --ccf-gov-msg-type ballot --ccf-gov-msg-created_at `date -uIs` --ccf-gov-msg-proposal_id "b9fd680f8f9f01ae916bfffabc1ccef54ace4a9407a673e41f111c5c871559e1" --signing-cert ../keys/memberccf_cert.pem --signing-key ../keys/memberccf_privk.pem --content vote_accept.json | curl 'https://192.168.122.82:443/gov/proposals/b9fd680f8f9f01ae916bfffabc1ccef54ace4a9407a673e41f111c5c871559e1/ballots' --cacert ../run/service_cert.pem --key ../keys/memberccf_privk.pem --cert ../keys/memberccf_cert.pem --data-binary @- -H "content-type: application/cose"
```

This results on following error:
```
{"error":{"code":"InternalError","message":"Failed to apply(): TypeError: Passed 0 arguments but expected two\nTrace:     at transitionServiceToOpen (native)\n    at <anonymous> (public:ccf.gov.constitution[0]:456)\n    at apply (public:ccf.gov.constitution[0]:839)\n"}}
```
Is this where error occurs?
https://github.com/microsoft/CCF/blob/ccf-4.0.16/src/js/wrap.cpp#L777
Not familiar with integration details but seems like this wrapper is expecting 2 arguments? Can you please share some insights.

Follow up question:
- with 1.x version the utility python scripts had all helpers to programmatically invoke this step. Seems 4.x only exposes COSE.py and ledger helpers. Is the recommendation to sign COSE & invoke curl API? Are there other python helpers that could be taken advantage of? (eg., the CCF/test has good collection of helpers used by sandbox)

Thanks!

**Comment by bsenthilr** [comment]
Add user proposal and ballot worked with same vote_accept.json. So in general this validates member is active and can vote and promote with same ballot endpoint for different action.

```
./ccf_cose_sign1 --ccf-gov-msg-type ballot --ccf-gov-msg-created_at `date -uIs` --ccf-gov-msg-proposal_id "e1ea6d484d82b628e1c91430bcdc7aa36b686fbb9b22eaab6d96bf4255a2356d" --signing-cert ../keys/memberccf_cert.pem --signing-key ../keys/memberccf_privk.pem --content vote_accept.json | curl https://192.168.122.82:443/gov/proposals/e1ea6d484d82b628e1c91430bcdc7aa36b686fbb9b22eaab6d96bf4255a2356d/ballots --cacert ../run/service_cert.pem --data-binary @- -H "content-type: application/cose"

{"ballot_count":1,"proposal_id":"e1ea6d484d82b628e1c91430bcdc7aa36b686fbb9b22eaab6d96bf4255a2356d","proposer_id":"e7b9bf87a47e8581ad0441a83f74e50339111e070c01f433771119f15e38301a","state":"Accepted","votes":{"e7b9bf87a47e8581ad0441a83f74e50339111e070c01f433771119f15e38301a":true}}
```

**Reply by bsenthilr** [reply]
ok guessing my action.js is not what is expected. Believe the container has old actions.js.
Let me cross check and reopen if the issue lingers. Bet thats my problem.

**Comment by achamayou** [comment, marked as answer]
> Assuming its still possible to open network without any user added.

It is, there is no restriction in that regard.

> The Json body does not look right. The transition_service_to_open action does not have any arguments.

It does in recent versions, please see https://github.com/microsoft/CCF/blob/main/samples/constitutions/default/actions.js#L595.

> with 1.x version the utility python scripts had all helpers to programmatically invoke this step. Seems 4.x only exposes COSE.py and ledger helpers. Is the recommendation to sign COSE & invoke curl API?

Yes

> Are there other python helpers that could be taken advantage of? (eg., the CCF/test has good collection of helpers used by sandbox)

There are some jinja templates under https://github.com/microsoft/CCF/tree/da247938395fe5c8eae55d9556c89d99fe32f098/samples/templates, but no supported Python helpers. The test infra is internal, you are of course welcome to use it, but it is not a supported API surface and may change without notice.

> Believe the container has old actions.js.

Yes, there have been updates since 1.x, the release notes from major releases (X.0) should describe them, but the quickest is probably to diff the current sample constitution vs yours and see what has changed.

**Reply by bsenthilr** [reply]
Thanks a lot @achamayou. My issue was basically using stale 1.x action.js. 
Will adopt to jinja templates. Will need to adopt it anyways for config migration from toml -> json.
